hugr-passes/src/dataflow/machine.rs

```diff
@@ -0,0 +1,160 @@
+use std::collections::HashMap;
+
+use hugr_core::{HugrView, IncomingPort, Node, PortIndex, Wire};
+
+use super::{datalog::AscentProgram, AbstractValue, DFContext, PartialValue};
+
+/// Basic structure for performing an analysis. Usage:
+/// 1. Get a new instance via [Self::default()]
+/// 2. (Optionally / for tests) zero or more [Self::prepopulate_wire] with initial values
+/// 3. Exactly one [Self::run], with initial values for root inputs, to do the analysis
+/// 4. Results then available via [Self::read_out_wire]
+pub struct Machine<V: AbstractValue, C: DFContext<V>>(
+    AscentProgram<V, C>,
+    Option<HashMap<Wire, PartialValue<V>>>,
+);
+
+/// derived-Default requires the context to be Defaultable, which is unnecessary
+impl<V: AbstractValue, C: DFContext<V>> Default for Machine<V, C> {
+    fn default() -> Self {
+        Self(Default::default(), None)
+    }
+}
+
+impl<V: AbstractValue, C: DFContext<V>> Machine<V, C> {
+    /// Provide initial values for some wires.
+    // Likely for test purposes only - should we make non-pub or #[cfg(test)] ?
+    pub fn prepopulate_wire(&mut self, h: &impl HugrView, wire: Wire, value: PartialValue<V>) {
+        assert!(self.1.is_none());
+        self.0.in_wire_value_proto.extend(
+            h.linked_inputs(wire.node(), wire.source())
+                .map(|(n, inp)| (n, inp, value.clone())),
+        );
+    }
+
+    /// Run the analysis (iterate until a lattice fixpoint is reached),
+    /// given initial values for some of the root node inputs.
+    /// (Note that `in_values` will not be useful for `Case` or `DFB`-rooted Hugrs,
+    /// but should handle other containers.)
+    /// The context passed in allows interpretation of leaf operations.
+    ///
+    /// # Panics
+    ///
+    /// If this Machine has been run already.
+    ///
+    pub fn run(
```

Comment:
Perhaps `run` should consume the Machine and return the out_wires, perhaps even in  a tuple with the `AscentProgram` (which could then be discarded, or inspected, but not run via Machine - perhaps manually if one wanted) ?

Reply:
(done)